hal/interfaces/FlashHeterogeneous.cpp

```diff
@@ -0,0 +1,53 @@
+#include "hal/interfaces/FlashHeterogeneous.hpp"
+#include <cstdint>
+
+namespace hal
+{
+    template<class T>
+    FlashHeterogeneousBase<T>::FlashHeterogeneousBase(infra::MemoryRange<T> sectors)
+        : sectors(sectors)
+    {}
+
+    template<class T>
+    T FlashHeterogeneousBase<T>::NumberOfSectors() const
+    {
+        return static_cast<T>(sectors.size());
+    }
+
+    template<class T>
+    uint32_t FlashHeterogeneousBase<T>::SizeOfSector(T sectorIndex) const
+    {
+        return static_cast<uint32_t>(sectors[sectorIndex]);
+    }
+
+    template<class T>
+    T FlashHeterogeneousBase<T>::SectorOfAddress(T address) const
+    {
+        T totalSize = 0;
+        for (T sector = 0; sector != sectors.size(); ++sector)
+        {
+            totalSize += sectors[sector];
+            if (address < totalSize)
+                return sector;
+        }
+
+        assert(address == totalSize);
+        return static_cast<T>(sectors.size() - 1);
```

Comment:
If address == totalSize, then the address is one past the last byte of flash, so the sector should be one past the last sector as well
```suggestion
        return static_cast<T>(sectors.size());
```